RFC/rfcBBL209/README.md

```diff
@@ -0,0 +1,104 @@
+# RFC|BB|L2-08: Handle Arbitrary Block Sizes
+
+* Status: `Brainstorm`
+
+## Abstract
+
+This RFC proposes adding a new type of data exchange to Bitswap for handling blocks of data arbitrarily larger than the 1MiB limit by using the features of common hash functions that allow for pausing and then resuming the hashes of large objects.
+
+## Shortcomings
+
+Bitswap has a maximum block size of 1MiB which means that it cannot transfer all forms of content addressed data. A prominent example of this is Git repos which even though they can be represented as a content addressed IPLD graph cannot necessarily be transferred over Bitswap if any of the objects in the repo exceed 1MiB.
+
+## Description
+
+The major hash functions work by taking some data `D` chunking it up into `n` pieces `P_0...P_n-1` then they modify an internal state `S` by loading pieces into the hash function in some way. This means that there are points in the hash function where we can pause processing and get the state of the hash function so far. Bitswap can utilize this state to effectively break up large blocks into smaller ones.
```

Comment:
Note that breaking blocks on hash function chunk boundaries means breaking rabin (or any other content based) chunking, which is essential for efficient content deduplication. For this to support arbitrary content boundaries for IPFS blocks, the hash state will need to include more than just the accumulated hash value; it will also need the trailing partial hash chunk data.

Reply:
> the hash state will need to include more than just the accumulated hash value; it will also need the trailing partial hash chunk data.

Can you elaborate more on what you were thinking here? What do you mean by accumulated hash value (my guess is the internal State) and the trailing partial hash (not sure what this means).

I'm thinking there are three different chunkers involved here:
1. The SHA-256 hash chunker which uses 64 byte boundaries
2. The exchange layer (i.e. Bitswap) chunker which emits the intermediate state `S` along the way, which has the restriction of every state being on a 64 byte boundary (e.g. we could use a fixed 256KiB, or a more complex function as long as it always ended on a 64B boundary)
3. The data storage layer where the data could be stored in any chunking fashion we want although we should store the exchange layer mappings for reuse when people download from us in the future

IIUC the thing that you're getting at here is that it's unfortunate that if Rabin would've saved us bandwidth here that it's only saving us disk storage because of the restrictions at the exchange layer.

----

I think there's potentially a way around this (which has tradeoffs) by allowing more data into the manifest. For example, when B responds to A giving them a manifest of blocks to download in addition to giving them a list of the intermediate States B could also send more information about the data corresponding to those states.

For example, B could send `(State, large block start index, large block end index, []ConsecutiveSubBlock)` where `ConsecutiveSubBlock = (SubBlockMultiHash, subblock start index, subblock end index)` and in this way A could decide whether to get the blocks corresponding to the State transition by either asking about bytes in the full block or by asking for the ConsecutiveSubBlocks and then using some of the bytes. This would allow B to tell A about a deduplicating chunking scheme they could use, but A 1) wouldn't be obligated to use it when downloading data 2) wouldn't be obligated to use it in their datastore.

WDYT?

Reply:
> > the hash state will need to include more than just the accumulated hash value; it will also need the trailing partial hash chunk data.
> 
> Can you elaborate more on what you were thinking here? What do you mean by accumulated hash value (my guess is the internal State) and the trailing partial hash (not sure what this means).

If you chunk on arbitrary byte boundaries, then the full "state" needed to resume hashing needs to include the tail data past the last 64B boundary that has not yet been included into the hash-chunker. It means the full state needed is a little larger.

> I'm thinking there are three different chunkers involved here:
> 
> 1. The SHA-256 hash chunker which uses 64 byte boundaries

Note this can vary depending on the hash function used. I think some have >64 byte hash-chunkers.

> 2. The exchange layer (i.e. Bitswap) chunker which emits the intermediate state `S` along the way, which has the restriction of every state being on a 64 byte boundary (e.g. we could use a fixed 256KiB, or a more complex function as long as it always ended on a 64B boundary)

This is the blocker; rabin chunkers must chunk on arbitrary byte boundaries and have variable sized chunks to work properly. The classic example is the large file modified by adding a single byte at the front; despite everything except the first byte being duplicated, chunkers that can't re-align on the 1-byte offset will fail to de-duplicate anything. Any chunker constrained to 64 byte boundaries will fail to find duplicates after any insert/delete that is not a nice multiple of 64 bytes.

> 3. The data storage layer where the data could be stored in any chunking fashion we want although we should store the exchange layer mappings for reuse when people download from us in the future

This bit is internal client level and doesn't really matter from an API point of view. The important bit is 2.

> IIUC the thing that you're getting at here is that it's unfortunate that if Rabin would've saved us bandwidth here that it's only saving us disk storage because of the restrictions at the exchange layer

Rabin can't be used at all if the chunks are constrained to 64 Byte boundaries. Without using some kind of content-based chunker like rabin, you get no de-duplication at all between slightly mutated data within files unless the data mutations are constrained to multiples of your block size; so no shifting of data except by multiples of the block size.

> I think there's potentially a way around this (which has tradeoffs) by allowing more data into the manifest. For example, when B responds to A giving them a manifest of blocks to download in addition to giving them a list of the intermediate States B could also send more information about the data corresponding to those states.

> For example, B could send `(State, large block start index, large block end index, []ConsecutiveSubBlock)` where `ConsecutiveSubBlock = (SubBlockMultiHash, subblock start index, subblock end index)` and in this way A could decide whether to get the blocks corresponding to the State transition by either asking about bytes in the full block or by asking for the ConsecutiveSubBlocks and then using some of the bytes. This would allow B to tell A about a deduplicating chunking scheme they could use, but A 1) wouldn't be obligated to use it when downloading data 2) wouldn't be obligated to use it in their datastore.

I don't think I fully understand this. Are you effectively creating virtual-large-blocks for each state transition out of a list of ranged sub-blocks in your manifest? And this is so that the sub-blocks can be created using a de-duplicating chunker, while the virtual-large-blocks can be 64byte aligned?

If yes, this sounds very similar to the "virtual DAG" idea I was proposing you could do using CID-aliases in the DHT. Note that you would not need to have a list of ConsecutiveSubBlocks, you could just have a CID+offset+length where the CID could be a reference to the merkle-dag node that contains the whole range. You can walk the DAG to find the leaf nodes needed from that.

Reply:
If you want a overview of how chunking/encoding/etc affect data deduplication, I think I've summarized everything about it here;

https://discuss.ipfs.io/t/draft-common-bytes-standard-for-data-deduplication/6813/10?u=dbaarda

Reply:
Sorry this is a little long, especially since I think we're basically on the same page but figured it'd help to be more explicit.

> If you chunk on arbitrary byte boundaries, then the full "state" needed to resume hashing needs to include the tail data past the last 64B boundary that has not yet been included into the hash-chunker. It means the full state needed is a little larger.
> ... 
> I don't think I fully understand this.

So I think what we're describing is just two ways of getting at the same idea. We're in agreement that the exchange layer chunker, which is verifiable, cannot just send over data-layer chunker information (e.g. rabin) since they won't line up on the same boundaries.

Overall we need something that a client who has `State_i` can receive `(State_i-1, bytes)`  and can verifiably check that the given state transition is valid.

IIUC the difference is:

### Your approach

I was having some difficulty describing what you were getting at here, but I think you were leaning towards:

Define transitions from `(State_i-1, bytes) -> State_i` as `(SHA2_State_with_extra_data_i-1, CIDs for data layer chunking blocks (e.g. rabin))` to `SHA2_State_with_extra_data_i` where the extra data might include some bytes left over in between rounds of SHA2 chunking.

### My approach

Define transitions from  `(State_i-1, bytes) -> State_i` as `(SHA2_State_i-1, []WayOfDescribingBytes) -> SHA2_State_i-1)` where I've listed two implementations of `WayOfDescribingBytes`.

Below we have three peers: Alice, the client. Bob the server that sends the manifest and/or data. Charlie another server that can send a manifest and/or data.

1. Byte offsets within the large block itself (works, but not great for deduplication)
2. Multihashes (since the IPLD codecs aren't required) of the blocks Bob has used locally to chunk the data (e.g. rabin) along with the start offset of the first block and end offset of the last block. For example, `[(MH1, start at byte 55), MH2, (MH3, end at byte 12)]`. Note: In one of your other comments I think you allude to allowing for graphs here instead of just a list of blocks, IMO that seems like an excessive level of indirection since we already know what we want.

Both of these approaches describe a set of bytes and have their own advantages/disadvantages:
1. Advantage: Works even if I ask Charlie for parts of the large block and Charlie has used a different chunker than Bob (e.g. buzzhash or fixed size). Disadvantage: Wastes bandwidth if Charlie and Bob used the same chunker, or if Alice had previously downloaded a large block (e.g. a file) that utilizes the data chunks in from how Bob has chunked up the data. 
2.Advantage: If people are using the same chunker I get deduplication benefits. Disadvantage: Fails completely if someone has the same large block, but chunked up differently

> Are you effectively creating virtual-large-blocks for each state transition out of a list of ranged sub-blocks in your manifest?

Yes, although the "large" in this case isn't very big since the sum of the sizes should be less than the recommended Bitswap block size (i.e. 1MiB). We're just accounting for if Rabin chunking gives us 10 chunks that contain that 1MiB of data.

This isn't perfect, for example if the data is a 1TiB file that's expressible using a rabin chunker in 1MiB we'll send metadata corresponding to a 1TiB file since we want to account for someone using a different chunker. However, as the manifests should be very small compared to the data and most real data likely won't have such extreme differences in chunking vs non-chunking size I suspect this is ok.

> And this is so that the sub-blocks can be created using a de-duplicating chunker, while the virtual-large-blocks can be 64byte aligned?

Yes, and also handle the case where people have used different chunkers for the same data.

> If you want a overview of how chunking/encoding/etc affect data deduplication

That's a great post 👍

Reply:
This is a long comment thread, but I just want to add that we get common-prefix deduplication of stuff with the linear hash for free.